SimG4CMS/Forward/src/BscNumberingScheme.cc

```diff
@@ -11,93 +11,94 @@
 #include "CLHEP/Units/GlobalSystemOfUnits.h"
 #include "globals.hh"
 
-BscNumberingScheme::BscNumberingScheme() { LogDebug("BscSim") << " Creating BscNumberingScheme"; }
-
-int BscNumberingScheme::detectorLevel(const G4Step* aStep) const {
-  //Find number of levels
-  const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
-  return (touch) ? ((touch->GetHistoryDepth()) + 1) : 0;
-}
-
-void BscNumberingScheme::detectorLevel(const G4Step* aStep, int& level, int* copyno, G4String* name) const {
-  //Get name and copy numbers
-  if (level > 0) {
+namespace {
+  int detectorLevel(const G4Step* aStep) {
+    //Find number of levels
     const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
-    for (int ii = 0; ii < level; ++ii) {
-      int i = level - ii - 1;
-      name[ii] = ForwardName::getName(touch->GetVolume(i)->GetName());
-      copyno[ii] = touch->GetReplicaNumber(i);
+    return (touch) ? ((touch->GetHistoryDepth()) + 1) : 0;
+  }
+
+  void detectorLevel(const G4Step* aStep, int level, std::vector<int>& copyno, std::vector<G4String>& name) {
+    //Get name and copy numbers
+    if (level > 0) {
+      copyno.reserve(level);
+      name.reserve(level);
+      const G4VTouchable* touch = aStep->GetPreStepPoint()->GetTouchable();
+      for (int ii = 0; ii < level; ++ii) {
+        int i = level - ii - 1;
+        name.push_back(ForwardName::getName(touch->GetVolume(i)->GetName()));
+        copyno.push_back(touch->GetReplicaNumber(i));
+      }
     }
   }
-}
+}  // namespace
 
-unsigned int BscNumberingScheme::getUnitID(const G4Step* aStep) const {
-  unsigned int intindex = 0;
-  int level = detectorLevel(aStep);
+namespace BscNumberingScheme {
+  unsigned int getUnitID(const G4Step* aStep) {
+    unsigned int intindex = 0;
+    int level = detectorLevel(aStep);
 
-  LogDebug("BscSim") << "BscNumberingScheme number of levels= " << level;
+    LogDebug("BscSim") << "BscNumberingScheme number of levels= " << level;
 
-  if (level > 0) {
-    int* copyno = new int[level];
-    G4String* name = new G4String[level];
-    detectorLevel(aStep, level, copyno, name);
+    if (level > 0) {
+      std::vector<int> copyno;
```

Comment:
why not just on the stack?
```
int copyno[level];
```
etc

Reply:
I think, compiler does not know what value may have "level".

Reply:
dynamic arrays are not part of the C++ standard so I do not use them.

Reply:
still they are fully supported by all compilers and executable is faster

Reply:
These detectors rarely hit by particles, so overhead is minimal. In truth, it is possible to change this code in a way that both arrays and vectors are not used. The first attempt in https://github.com/cms-sw/cmssw/pull/41942 was not successful in testing. So, I would left things as they are.